### PR TITLE
Strategy node - Make strategy less unreadable

### PR DESCRIPTION
**Pull request by SimLemay**

### CODE REVIEW

File: ai/Algorithm/Graph/Graph.py [hunk 1]
```diff
@@ -21,18 +21,18 @@ class Graph:
...
     def __init__(self):
         self.nodes = []
-        self.current_node = 0
+        self.current_node = None
 
     def get_current_tactic_name(self):
         """
         :return: Le nom de la tactique en cours, sous forme d'une chaîne de caractères.
         """
         if len(self.nodes) > 0:
```
Comment (sim0504) [inline ai/Algorithm/Graph/Graph.py]: if self.current_node is not None:

File: ai/Algorithm/Graph/Graph.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def get_current_tactic(self):
         :return: La tactique en cours.
         """
         if len(self.nodes) > 0:
```
Comment (sim0504) [inline ai/Algorithm/Graph/Graph.py]: if self.current_node is not None:

File: ai/STA/Strategy/indirect_free_kick.py [hunk 3]
```diff
@@ -27,12 +27,12 @@ def __init__(self, p_game_state):
...
+        self.create_node(Role.GOALKEEPER, GoalKeeper(self.game_state, goalkeeper, ourgoal))
 
         for index, player in role_by_robots:
             if player:
-                self.add_tactic(index, PositionForPass(self.game_state, player, auto_position=True))
-                self.add_tactic(index, GoKick(self.game_state, player, auto_update_target=True))
+                self.create_node(index, PositionForPass(self.game_state, player, auto_position=True))
+                self.create_node(index, GoKick(self.game_state, player, auto_update_target=True))
 
                 self.add_condition(index, 0, 1, partial(self.is_closest, player))
```
Comment (sim0504) [inline ai/STA/Strategy/indirect_free_kick.py]: need to be connect_to(...)

File: ai/STA/Strategy/offense.py [hunk 4]
```diff
@@ -27,16 +27,27 @@ def __init__(self, p_game_state):
...
             if player:
-                self.add_tactic(index, PositionForPass(self.game_state, player, auto_position=True))
-                self.add_tactic(index, GoKick(self.game_state, player, auto_update_target=True))
+#                self.create_node(index, PositionForPass(self.game_state, player, auto_position=True))
+#                self.create_node(index, GoKick(self.game_state, player, auto_update_target=True))
 
-                self.add_condition(index, 0, 1, partial(self.is_closest, player))
-                self.add_condition(index, 1, 0, partial(self.is_not_closest, player))
-                self.add_condition(index, 1, 1, partial(self.has_kicked, player))
+#                self.add_condition(index, 0, 1, partial(self.is_closest, player))
```
Comment (sim0504) [inline ai/STA/Strategy/offense.py]: remove comments

File: ai/STA/Strategy/offense_3v3.py [hunk 5]
```diff
@@ -32,13 +32,13 @@ def __init__(self, p_game_state):
...
 
         for index, player in role_by_robots:
             if player:
-                self.add_tactic(index, PositionForPass(self.game_state, player, auto_position=True,
+                self.create_node(index, PositionForPass(self.game_state, player, auto_position=True,
                                                        robots_in_formation=self.robots))
-                self.add_tactic(index, GoKick(self.game_state, player, target=self.theirgoal))
+                self.create_node(index, GoKick(self.game_state, player, target=self.theirgoal))
 
                 self.add_condition(index, 0, 1, partial(self.is_closest, player))
```
Comment (sim0504) [inline ai/STA/Strategy/offense_3v3.py]: connect_to

File: ai/STA/Strategy/passes_with_decisions.py [hunk 6]
```diff
@@ -26,9 +26,9 @@ def __init__(self, p_game_state):
...
-        self.add_tactic(Role.FIRST_ATTACK, PassToPlayer(self.game_state, self.game_state.get_player_by_role(Role.FIRST_ATTACK), target_id=self.game_state.get_player_by_role(Role.MIDDLE).id))
-        self.add_tactic(Role.FIRST_ATTACK, GoKick(self.game_state, self.game_state.get_player_by_role(Role.FIRST_ATTACK), self.goal))
+        self.create_node(Role.FIRST_ATTACK, PassToPlayer(self.game_state, self.game_state.get_player_by_role(Role.FIRST_ATTACK), target_id=self.game_state.get_player_by_role(Role.SECOND_ATTACK).id))
+        self.create_node(Role.FIRST_ATTACK, PassToPlayer(self.game_state, self.game_state.get_player_by_role(Role.FIRST_ATTACK), target_id=self.game_state.get_player_by_role(Role.MIDDLE).id))
+        self.create_node(Role.FIRST_ATTACK, GoKick(self.game_state, self.game_state.get_player_by_role(Role.FIRST_ATTACK), self.goal))
 
         self.add_condition(Role.FIRST_ATTACK, 0, 1, partial(self.is_best_receiver, Role.SECOND_ATTACK))
         self.add_condition(Role.FIRST_ATTACK, 0, 2, partial(self.is_best_receiver, Role.MIDDLE))
```
Comment (sim0504) [inline ai/STA/Strategy/passes_with_decisions.py]: connect_to

File: tests/Algorithm/test_node.py [hunk 7]
```diff
@@ -26,8 +26,8 @@ def setUp(self):
         self.anotherTactic = TestNode._create_mock_tactic("Another command")
         self.node1 = Node(self.aTactic)
         self.node2 = Node(self.anotherTactic)
-        self.vertex1 = Vertex(0, foo)
-        self.vertex2 = Vertex(1, foo2)
+        self.vertex1 = Vertex(self.node1, return_true)
```
Comment (sim0504) [inline tests/Algorithm/test_node.py]: could be a `lambda: True`

**Comment by sim0504** [comment]
the add_condition  need to be refactor to connect_to(... when=)

File: ai/Algorithm/Graph/Graph.py [hunk 8]
```diff
@@ -21,18 +21,18 @@ class Graph:
         __str__: Retourne une représentation du graphe sous forme d'une chaîne de caractères.
     Attributs:
         nodes: Une liste des noeuds du graphe.
-        current_node: L'index du noeud courant dans la liste des noeuds.
+        current_node: Un pointeur vers le noeud courant dans la liste des noeuds.
```
Comment (PhiBabin) [inline ai/Algorithm/Graph/Graph.py]: Une référence vers

File: ai/Algorithm/Graph/Node.py [hunk 9]
```diff
@@ -41,17 +42,8 @@ def add_vertex(self, p_vertex):
...
-        Retire un vertex du noeud, spécifié par le numéro de son noeud d'arrivé. Si le noeud courant ne possède pas de
-        vertex vers le noeud spécifé, rien ne se passe.
-        :param p_ending: Un entier positif représentant le numéro du noeud d'arrivé du vertex à retirer.
-        """
-        assert isinstance(p_ending, int)
-        assert 0 <= p_ending
-        for i in range(len(self.vertices)):
-            if self.vertices[i].next_node == p_ending:
-                self.vertices.pop(i)
+    def connect_to(self, dst_node, *, when: Callable[..., bool]):
```
Comment (PhiBabin) [inline ai/Algorithm/Graph/Node.py]: Why the `*,` ?

File: ai/STA/Strategy/defense_wall.py [hunk 10]
```diff
@@ -27,14 +27,14 @@ def __init__(self, game_state: GameState, number_of_players: int = 4):
...
         self.robots = [player for _, player in role_by_robots if player is not None]
         for role, player in role_by_robots:
             if player:
-                self.add_tactic(role, AlignToDefenseWall(self.game_state, player, self.robots))
-                self.add_tactic(role, GoKick(self.game_state, player, target=self.theirgoal))
+                self.create_node(role, AlignToDefenseWall(self.game_state, player, self.robots))
+                self.create_node(role, GoKick(self.game_state, player, target=self.theirgoal))
 
                 self.add_condition(role, 0, 1, partial(self.is_closest, player))
                 self.add_condition(role, 1, 1, partial(self.is_closest, player))
```
Comment (PhiBabin) [inline ai/STA/Strategy/defense_wall.py]: You need to upgrade those add_condition to fit the connect_to() model we now have.

File: ai/STA/Strategy/direct_free_kick.py [hunk 11]
```diff
@@ -27,12 +27,12 @@ def __init__(self, p_game_state):
...
 
         for index, player in role_by_robots:
             if player:
-                self.add_tactic(index, PositionForPass(self.game_state, player, auto_position=True))
-                self.add_tactic(index, GoKick(self.game_state, player, auto_update_target=True))
+                self.create_node(index, PositionForPass(self.game_state, player, auto_position=True))
+                self.create_node(index, GoKick(self.game_state, player, auto_update_target=True))
 
                 self.add_condition(index, 0, 1, partial(self.is_closest, player))
                 self.add_condition(index, 1, 0, partial(self.is_not_closest, player))
```
Comment (PhiBabin) [inline ai/STA/Strategy/direct_free_kick.py]: Same thing here

File: ai/Algorithm/Graph/Node.py (hunk 9 above)
Comment (sim0504) [reply ai/Algorithm/Graph/Node.py]: It make sure you use when as a keyword argument

Comment (PhiBabin) [reply ai/Algorithm/Graph/Node.py]: The more you know

File: ai/Algorithm/Graph/Graph.py (hunk 1 above)
Comment (sim0504) [reply ai/Algorithm/Graph/Graph.py]: also, it could be a @property

**Comment by wonwon0** [comment]
Le nom de la branche: "make branches name more unconfusing"

File: ai/Algorithm/Graph/Graph.py [hunk 12]
```diff
@@ -53,67 +55,58 @@ def add_node(self, p_node):
...
-        assert isinstance(p_node_index, int)
-        assert 0 <= p_node_index < len(self.nodes)
-        for i in range(len(self.nodes)):
-            self.remove_vertex(i, p_node_index)
-        self.nodes.pop(p_node_index)
+        assert isinstance(dst_node, Node)
+        assert dst_node in self.nodes
+        for src_node in self.nodes:
+            self.remove_vertex(src_node, dst_node)
+        self.nodes.pop(dst_node)
```
Comment (MaximeGLegault) [inline ai/Algorithm/Graph/Graph.py]: Dubious use of pop methods. Also method is not used.

**Pull request merged**